src/runtime/rpc/network.cpp

```diff
@@ -272,18 +287,27 @@ void rpc_session::send_message(message_ex *msg)
     uint64_t sig;
     {
         utils::auto_lock<utils::ex_lock_nr> l(_lock);
-        msg->dl.insert_before(&_messages);
-        ++_message_count;
 
-        if (SS_CONNECTED == _connect_state && !_is_sending_next) {
-            _is_sending_next = true;
-            sig = _message_sent + 1;
-            unlink_message_for_send();
+        // Attention: here we only allow two cases to send message:
+        //  case 1: session's state is SS_CONNECTED
+        //  case 2: session is sending negotiation message
+        if (SS_CONNECTED == _connect_state || security::is_negotiation_message(msg->rpc_code())) {
+            msg->dl.insert_before(&_messages);
+            ++_message_count;
+
+            if (!_is_sending_next) {
+                _is_sending_next = true;
+                sig = _message_sent + 1;
+                unlink_message_for_send();
+            } else { // is under send msg, just return
+                return;
+            }
         } else {
+            dassert(_connect_state != SS_CONNECTED, "invalid session state");
```

Comment:
Use `dcheck_ne` (check not equal) would be better, then when check failed, you will know the exact value of  _connect_state which may help you to debug.
However, now `dcheck_ne` is lacked, you can add it by another PR.

Reply:
I will remove `dassert(_connect_state != SS_CONNECTED, "invalid session state");`, because `_connect_state != SS_CONNECTED` is always true in the else code branch